Make PropertyAndSetterInjection field sorting consistent

josephw opened the pull request:
#155 is caused by calling `Collections.sort` with a comparator that isn't always transitive, as in the case where sorting by name and sorting by type hierarchy imply different orders.

For example, with fields of type `[Iterable, Integer, Number]`, it gives `Iterable == Integer` and `Iterable == Number`, but also an inconsistent `Integer < Number`.

I'm not sure it's possible to sort consistently with a stateless `Comparator`, so this change switches to a `sortSuperTypesLast` method that first sorts by name and then shuffles any subclasses to after their superclasses. It's slower, but should be deterministic. The existing tests are preserved by leaving behind a dummy `Comparator` that uses the new sort method. New tests make sure that the sorted ordering doesn't depend on the original order.

Fixes #155.

bric3 commented:
Wow, nice PR, thanks for diving in the problem !

So I agree on the idea, but I've mixed feelings on having a custom sort there.
Also I agree speed is not the real issue here anyway, I'd prefer the algorithm to be deterministic.

bric3 commented on src/org/mockito/internal/configuration/injection/PropertyAndSetterInjection.java:
```diff
@@ -125,27 +124,45 @@ private boolean injectMockCandidatesOnFields(Set<Object> mocks, Object instance,
+                i++;
+            } else {
+                fields.remove(i);
+                fields.add(newPos, f);
```
I'm not comfortable to have a custom sort algorithm I would prefer the use of comparator, it could be stateful of course.

bric3 replied:
Also I didn't thought about that, but changing the type to a `TreeSet` may be interesting.

bric3 commented on test/org/mockito/internal/configuration/injection/PropertyAndSetterInjectionTest.java:
```diff
@@ -0,0 +1,151 @@
+        }
+
+        assertThat(results).hasSize(1);
+    }
```
nice test !

bric3 commented:
Finally I like the comment in the commit messages, though it would be better if the last 2 commit were amended properly :)

bric3 commented:
Overall it's really nice work :sparkles: 

I you think it's possible to rewrite with a comparator and rewrite a clean history I'll merge _asap_ :)

josephw commented:
Thanks; I've rebased to clean up history.

Using `Comparator` is trickier. I'd really prefer to do that, and I spent some time thinking about what state could be calculated to allow pairwise comparison. However, this is essentially a topological sort, and I can't think of a way to do this short of using a custom sort and then preparing a `Comparator` that checks indexes in that version.

Actually, checking now, that's exactly the fallback suggested in [this Stackoverflow answer](http://stackoverflow.com/questions/28099307/how-to-write-a-transitive-comparator-when-equality-implies-order-doesnt-matt):

> The only way I see that it would be possible to do using a Comparator would be to first do a topological sort, and then using the indexes of the objects in this sorting as keys when implementing the comparator.

I'd be happy to add any more tests, or change the style to increase confidence that it's a safe implementation.

bric3 commented:
Mmmmh OK, it would definitely be better to have the comparator, but that seems a bit complex as you said. So I propose 2 things then :

1. Without comparator it would be better to externalize the sorting in some class instead of a method, could be used like 

    ```java
    sorted = new org.mockito.internal.util.reflection.TopologySorter().sort(...)
    ```

2. Thanks for the amend, I propose that the proper commits starts `Fixes #155, ` like you did as github can track such keyword and issue reference.

Sorry to be nitpicky, but as soon as the PR is merged there will be an automated release :)
Also tests are fine.

Thanks for the feedback, and the interesting links :)

bric3 commented:
:star2: 

Thanks for the modifications @josephw 
Perfect PR :)

bric3 commented:
By the way do you have any remarks on the contributing guide (missing info, style guideline, git stuff, etc..)

josephw commented:
It's great that you've got a contribution guide. I think a couple of the things I missed are mentioned by it but, I might over-emphasise:

>  At least one commit message in the PR contains "Fixes #id"

and suggest a template commit message.

> Clean commit history

You could make the history rewriting explicit here. Something like 'We prefer a clean commit history, rebased and squashed if necessary, so that each commit clearly changes one things and there are no extraneous fix-ups'?

Thanks for the feedback -- everything else seemed specific enough that a review was the only way to pick it up.

bric3 commented:
@josephw Ok thanks for the feedback too
